opt: use existing partial indexes to recommend new partial indexes

HasithDeAlwis opened the pull request:
This PR completes partial index recommendation support in CockroachDB's index recommendation engine, enabling the system to suggest indexes with WHERE clauses when they would improve query performance. The implementation has the following logical steps
1. Candidate generation now creates `IndexCandidates` with predicate information from existing partial indexes
2. Create hypothetical index candidates with a `predicate` field populated
3. In `rec.go` it now do NOT skip over partial indexes as recommendations or as possible candidates to replace
4. SQL generation properly formats CREATE INDEX statements with WHERE clauses using parsed predicate expressions. 

Closes #143425

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154564)
<!-- Reviewable:end -->

mgartner reviewed (commented):
Thanks for the contribution! I did a first pass and left a few comments and questions.

mgartner commented on pkg/sql/opt/indexrec/candidate.go:
```diff
@@ -50,12 +50,62 @@ import (
+
+			predicateInfos := tablesToPredicates[t]
+			for _, predicateInfo := range predicateInfos {
+				ics.finalCandidates[t] = append(ics.finalCandidates[t], IndexCandidates{
```
My understanding of #143425 is that we only want to suggest adding key or storing columns to existing partial indexes, not adding any partial index predicate to any possible new index.

For example:

```sql
CREATE TABLE t (
  i INT,
  s STRING,
  f FLOAT,
  INDEX (s) WHERE i > 0
)

SELECT * FROM t WHERE f = 1.0 AND i > 0
```

As far as I understand we don't want to recommend `CREATE INDEX ON t(f) WHERE i > 0`—that's too big of an assumption to make without the context of the application logic.

It looks like your current implementation could recommend that—but I'm not sure. Can you add some tests like this?

HasithDeAlwis commented on pkg/sql/opt/indexrec/candidate.go:
```diff
@@ -67,6 +117,38 @@ type indexCandidateSet struct {
+
+type IndexCandidates struct {
+	columns       []cat.IndexColumn
+	predicateInfo PredicateInfo
```
Ah, thank you, this was supposed to get cleaned up after I refactored some logic :+1:

HasithDeAlwis committed:
sql/opt: create infrastructure to handle to generate index with predicates

chore: squash

HasithDeAlwis committed:
sql/opt: add functionality to preserve `partialIndexPredicates` on table metadata when updating

HasithDeAlwis committed:
sql/opt: recommend partial indicies on performance gains

Release note(sql-change): The index recommendation engine now suggests new partial indexes based on a table's existing partial index predicates. This helps users discover performance improvements automatically without manually reasoning about complex predicates.

HasithDeAlwis reviewed (commented):
@mgartner did the changes you asked for, updating candidate generation and test case creation!

mgartner reviewed (changes requested):
Sorry for the delay. This is getting simpler, but I don't think we're on the same page about how this feature should work. Take a look at the comment I left below and please let me know if you have any questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-:-OdjLn3pAsXwEB1DulJW:b-hatem)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @HasithDeAlwis)

___
*[`-- commits` line 10 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-OdjHehGD-dmXDK_6_TJ:-OdjHehGD-dmXDK_6_TK:b-u48dvx):*
nit: wrap git commits at 80 characters
___
*[`pkg/sql/opt/indexrec/hypothetical_index.go` line 192 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-OdjIErq-KaNqpCStgfa:-OdjIErq-KaNqpCStgfb:b-2iccwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6a9d849d9dd338929429f8c849d7bc86ecf82d/pkg/sql/opt/indexrec/hypothetical_index.go#L192)):*
> ```Go
> 		return "", false
> 	}
> 	return hi.predicate, true
> ```

nit: this can be simplified to:

```go
return hi.predicate, hi.predicate != ""
```
___
*[`pkg/sql/opt/indexrec/candidate.go` line 104 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-OdjH9c_8layGpTvyghZ:-OdjH9c_8layGpTvygh_:bjirsuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecf077f9586a03a6bb58399301687d6fccf8f254/pkg/sql/opt/indexrec/candidate.go#L104)):*
> ```Go
> 
> // A helper function that checks a predicate against the columns of an index.
> // If every column used in the predicate is present in the index columns, it is a valid index and should be evaluated as a candidate.
> ```

I don't understand why this criteria is important. Can you explain more? If we have a candidate index like `INDEX (a)`, and an existing index like `INDEX (b) WHERE c > 0`, don't we want to consider the candidate `INDEX (a) WHERE c > 0`? Wouldn't this logic prevent that candidate from being added?

nit: wrap comments at 80 characters
___
*[`pkg/sql/opt/indexrec/candidate.go` line 74 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-OdjHvcu5HQGZwvMtKp3:-OdjHvcu5HQGZwvMtKp4:b1024ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6a9d849d9dd338929429f8c849d7bc86ecf82d/pkg/sql/opt/indexrec/candidate.go#L74)):*
> ```Go
> 
> // Wraper type that can maintain the predicate associated with an index candidate.
> type IndexCandidates struct {
> ```

nit: I think this should be called `IndexCandidate` instead, since it represents a single candidate, right?
___
*[`pkg/sql/opt/indexrec/candidate.go` line 87 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-OdjI8pY-ydcLAUtn4hv:-OdjI8pY-ydcLAUtn4hw:b-dbn3e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6a9d849d9dd338929429f8c849d7bc86ecf82d/pkg/sql/opt/indexrec/candidate.go#L87)):*
> ```Go
> }
> 
> // Predicate returns the predicate string.
> ```

nit: this comment is in the wrong place.
___
*[`pkg/sql/opt/indexrec/hypothetical_table_test.go` line 28 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-OdjITm24s1nEN_IGikJ:-OdjITm24s1nEN_IGikK:b-pl6okf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6a9d849d9dd338929429f8c849d7bc86ecf82d/pkg/sql/opt/indexrec/hypothetical_table_test.go#L28)):*
> ```Go
> 			candidates[i] = IndexCandidates{
> 				columns:   indexColumns,
> 				predicate: "",
> ```

nit: this function has a confusing name if it never sets the predicate. 


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/indexrec/candidate.go` line 104 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/154564#-OdjH9c_8layGpTvyghZ:-OdjM7mq9_ghoCGDUuu8:b64ig3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecf077f9586a03a6bb58399301687d6fccf8f254/pkg/sql/opt/indexrec/candidate.go#L104)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't understand why this criteria is important. Can you explain more? If we have a candidate index like `INDEX (a)`, and an existing index like `INDEX (b) WHERE c > 0`, don't we want to consider the candidate `INDEX (a) WHERE c > 0`? Wouldn't this logic prevent that candidate from being added?

nit: wrap comments at 80 characters
</blockquote></details>

Disregard my example above. Here is a correct one:

If we have an existing index `INDEX (a) WHERE b > 0` and we have an index candidate `INDEX (a, c)`, then we also want to try the candidate `IDNEX (a, c) WHERE b > 0`. I think this logic would prevent that, wouldn't it?


<!-- Sent from Reviewable.io -->

HasithDeAlwis commented:
> If we have an existing index `INDEX (a) WHERE b > 0` and we have an index candidate `INDEX (a, c)`, then we also want to try the candidate `INDEX (a, c) WHERE b > 0`. I think this logic would prevent that, wouldn't it?

Yeah, you're right this logic would prevent that, I misunderstood you're original feedback! 
Looking at what you commented above, I have a good idea of what we actually want from this feature. Will ping again once I have the changes needed
